frame/society/src/lib.rs

```diff
@@ -417,108 +365,320 @@ impl<AccountId: PartialEq, Balance> BidKind<AccountId, Balance> {
+	pub(super) type Defender<T: Config<I>, I: 'static = ()> = StorageValue<_, T::AccountId>;
+
+	/// Votes for the defender.
+	#[pallet::storage]
+	pub(super) type DefenderVotes<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Twox64Concat, T::AccountId, Vote>;
+
+	/// The max number of members for the society at one time.
+	#[pallet::storage]
+	#[pallet::getter(fn max_members)]
+	pub(super) type MaxMembers<T: Config<I>, I: 'static = ()> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let mut members = vec![];
 
-		/// The amount of the unpaid reward that gets deducted in the case that either a skeptic
-		/// doesn't vote or someone votes in the wrong way.
-		const WrongSideDeduction: BalanceOf<T, I> = T::WrongSideDeduction::get();
+			let mut weight = 0;
+			let weights = T::BlockWeights::get();
 
-		/// The number of times a member may vote the wrong way (or not at all, when they are a skeptic)
-		/// before they become suspended.
-		const MaxStrikes: u32 = T::MaxStrikes::get();
+			// Run a candidate/membership rotation
+			if (n % T::RotationPeriod::get()).is_zero() {
+				members = <Members<T, I>>::get();
+				Self::rotate_period(&mut members);
 
-		/// The amount of incentive paid within each period. Doesn't include VoterTip.
-		const PeriodSpend: BalanceOf<T, I> = T::PeriodSpend::get();
+				weight += weights.max_block / 20;
+			}
 
-		/// The number of blocks between candidate/membership rotation periods.
-		const RotationPeriod: T::BlockNumber = T::RotationPeriod::get();
+			// Run a challenge rotation
+			if (n % T::ChallengePeriod::get()).is_zero() {
+				// Only read members if not already read.
+				if members.is_empty() {
+					members = <Members<T, I>>::get();
+				}
+				Self::rotate_challenge(&mut members);
 
-		/// The number of blocks between membership challenges.
-		const ChallengePeriod: T::BlockNumber = T::ChallengePeriod::get();
+				weight += weights.max_block / 20;
+			}
 
-		/// The societies's module id
-		const PalletId: PalletId = T::PalletId::get();
+			weight
+		}
+	}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config<I>, I: 'static = ()> {
+		pub pot: BalanceOf<T, I>,
+		pub members: Vec<T::AccountId>,
+		pub max_members: u32,
+	}
 
-		/// Maximum candidate intake per round.
-		const MaxCandidateIntake: u32 = T::MaxCandidateIntake::get();
+	#[cfg(feature = "std")]
+	impl<T: Config<I>, I: 'static> Default for GenesisConfig<T, I> {
+		fn default() -> Self {
+			Self {
+				pot: Default::default(),
+				members: Default::default(),
+				max_members: Default::default(),
+			}
+		}
+	}
 
-		// Used for handling module events.
-		fn deposit_event() = default;
+	#[pallet::genesis_build]
+	impl<T: Config<I>, I: 'static> GenesisBuild<T, I> for GenesisConfig<T, I> {
+		fn build(&self) {
+			Pot::<T, I>::put(self.pot);
+			MaxMembers::<T, I>::put(self.max_members);
```

Comment:
Should we add a warning if `self.max_members >= self.members.len()`? cc @thiolliere

Reply:
yes and we should even panic during genesis, this will happen when we implement the bounding of storages.
But as it wasn't done before I think we shouldn't block on this.